libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c

```diff
@@ -376,166 +278,591 @@ CK_RV prvMbedTLS_Initialize( void )
         {
-            memset( pxSessionObj, 0, sizeof( P11Session_t ) );
-        }
+            if( 0 == memcmp( xLabel.pValue, pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS, xLabelLength ) )
+            {
+                prvFindObjectInListByLabel( ( uint8_t * ) pkcs11configLABEL_DEVICE_PUBLIC_KEY_FOR_TLS, strlen( ( char * ) pkcs11configLABEL_DEVICE_PUBLIC_KEY_FOR_TLS ), &xPalHandle, &xAppHandle2 );
 
-        pxSessionObj->xSignMutex = xSemaphoreCreateMutex();
+                if( xPalHandle != CK_INVALID_HANDLE )
+                {
+                    xResult = prvDeleteObjectFromList( xAppHandle2 );
+                }
+            }
+            else if( 0 == memcmp( xLabel.pValue, pkcs11configLABEL_DEVICE_PUBLIC_KEY_FOR_TLS, xLabelLength ) )
+            {
+                prvFindObjectInListByLabel( ( uint8_t * ) pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS, strlen( ( char * ) pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS ), &xPalHandle, &xAppHandle2 );
 
-        if( NULL == pxSessionObj->xSignMutex )
+                if( xPalHandle != CK_INVALID_HANDLE )
+                {
+                    xResult = prvDeleteObjectFromList( xAppHandle2 );
+                }
+            }
+
+            xResult = prvDeleteObjectFromList( xAppHandle );
+        }
+
+        if( xFreeMemory == CK_TRUE )
         {
-            xResult = CKR_HOST_MEMORY;
+            PKCS11_PAL_GetObjectValueCleanup( pxObject, ulObjectLength );
         }
 
-        pxSessionObj->xVerifyMutex = xSemaphoreCreateMutex();
+        return xResult;
+    }
+
+#endif /* if ( pkcs11configPAL_DESTROY_SUPPORTED != 1 ) */
+
+#if ( pkcs11configJITP_CODEVERIFY_ROOT_CERT_SUPPORTED != 1 )
+
+/* Returns True if the object is not stored in NVM & must be looked up in header file.
+ * Returns fals if object is an NVM supported object. */
+
+    BaseType_t xIsObjectWithNoNvmStorage( uint8_t * pucLabel,
+                                          size_t xLength,
+                                          uint8_t ** ppucCertData )
+    {
+        BaseType_t xResult = CK_TRUE;
+
+        if( 0 == memcmp( pucLabel, pkcs11configLABEL_JITP_CERTIFICATE, strlen( ( char * ) pkcs11configLABEL_JITP_CERTIFICATE ) ) )
+        {
+            if( NULL != keyJITR_DEVICE_CERTIFICATE_AUTHORITY_PEM )
+            {
+                *ppucCertData = ( uint8_t * ) keyJITR_DEVICE_CERTIFICATE_AUTHORITY_PEM;
+            }
+            else
+            {
+                PKCS11_PRINT( ( "ERROR: JITP Certificate not specified.\r\n" ) );
+            }
+        }
+        else if( 0 == memcmp( pucLabel, pkcs11configLABEL_ROOT_CERTIFICATE, strlen( ( char * ) pkcs11configLABEL_ROOT_CERTIFICATE ) ) )
+        {
+            /* Use either Verisign or Starfield root CA,
+             * depending on whether this is an ATS endpoint. */
+            if( strstr( clientcredentialMQTT_BROKER_ENDPOINT, "-ats.iot" ) == NULL )
+            {
+                *ppucCertData = ( uint8_t * ) tlsVERISIGN_ROOT_CERTIFICATE_PEM;
```

Comment:
Ah, I understand now that the ATS certs are cross-signed by Starfield.  Still, it seems like we should be using Amazon's own root certs.

Reply:
yes sorry I should have mentioned that.  ATM we cannot always use the Amazon Root Certificate because some platforms do not trust Amazon's Root CA, but do have built in trust for the Starfield one.  I think updating device SDKs in our repo might take care of that issue, but until this has done I will leave as is.